Mounting sources unification

VladimirSlavik opened the pull request:
This makes sources that mount share common ancestors.

Pending:

- [x] Pylint has a problem because I have a subclass that is meant to be abstract, and does not implement some stuff. ~~Not sure what to do with that.~~ It's broken so adding ignores.
- [x] Naming and locations of classes need feedback.
- [x] Tests for the base classes.
- [x] What to do about ancestry of the HDISO stuff. -> Another PR.
- [x] Generating the mount points

VladimirSlavik commented:
Pylint has a problem because I have a subclass that is meant to be abstract, and does not implement some stuff. Not sure what to do with that.

poncovka commented on tests/nosetests/pyanaconda_tests/module_source_hmc_test.py:
```diff
@@ -58,7 +57,9 @@ def ready_state_test(self, ismount):
         self.assertTrue(self.module.is_ready())
-        ismount.assert_called_once_with(INSTALL_TREE)
+        ismount.assert_called_once_with(
+            "/mnt/anaconda-sources/hmc-{:0>4}".format(self.module.counter - 1)
```
Please, use `self.module.mount_point`. Same in other tests.

poncovka commented on tests/nosetests/pyanaconda_tests/module_source_nfs_test.py:
```diff
@@ -115,9 +114,12 @@ def ready_state_test(self, ismount):
 
         self.assertTrue(self.source_module.is_ready())
```
Btw. I think that you can remove most of the tests for the `is_ready` method and keep only one.

VladimirSlavik replied:
Haha, I really am braindead :) Yes, the tests need way more changes, namely for the base classes. Sorry.

VladimirSlavik replied:
I think I have addressed that by now...

poncovka commented on pyanaconda/modules/payloads/source/harddrive/harddrive.py:
```diff
@@ -43,8 +41,8 @@ def __init__(self):
         self._install_tree_path = ""
-        self._device_mount = INSTALL_TREE + "_device"
-        self._iso_mount = INSTALL_TREE + "_iso"
+        self._device_mount = self._mount_point + "-device"
```
Maybe I would override the `mount_point` property to return the device mount point. Then you can use the `super().is_ready` method.

jkonecny12 replied:
Not sure if this suggestion is not bending the purpose of the `MountingSourceBase` class too much.

VladimirSlavik replied:
Maybe the real problem here is whether harddrive belongs with other mountables or not?

In a very broad sense it does fit: it mounts, and wants a number. Then again it has multiple mount points, actually a variable number at that, so the actual functionality is different.

poncovka replied:
If the `mount_point` property cannot return a real mount point, then `HardDriveSourceModule` shouldn't be a subclass of `MountingSourceBase`. However, we need to be able to generate the "base" mount point. I still think that there should be a separate class called `MountPointGenerator` with one class method `create_mount_point` that will return a unique mount point. There might be more cases like this one in the future.

poncovka commented on pyanaconda/modules/payloads/source/harddrive/harddrive.py:
```diff
@@ -43,8 +41,8 @@ def __init__(self):
-        self._device_mount = INSTALL_TREE + "_device"
-        self._iso_mount = INSTALL_TREE + "_iso"
+        self._device_mount = self._mount_point + "-device"
+        self._iso_mount = self._mount_point + "-iso"
```
Maybe, you could replace `TearDownHardDriveSourceTask` with two instances of `TearDownMountTask`. It would be good to also use `SetUpMountTask` to check the mount points.

VladimirSlavik replied:
The former is another facepalm moment for me. The latter, I am not sure. I'll try to remember and bring it up at the next opportunity.

VladimirSlavik replied:
Haha, another facepalm - I finally got what you meant by the other half. Check implemented in-place now, discussion still needed. I don't see how a class that assumes exactly one of something can be subclassed easily into handling more than one of that. I'd need to override all of the ancestor code anyway.

poncovka replied:
This solution seems to be fine.

poncovka commented:
About the pylint issues, maybe try to remove ABCMeta and ABC from `PayloadSourceBase` and `MountingSourceBase`. They already inherit from ABC. For example, `AbstractTask` seems to work this way.

VladimirSlavik commented:
> About the pylint issues (...)

Apparently pylint puts too little work into deciding what is abstract so I just added an ignore for that in the class.

jkonecny12 commented on pyanaconda/modules/payloads/source/source_base.py:
```diff
@@ -84,3 +89,50 @@ def tear_down_with_tasks(self):
+        self._mount_point = self._generate_mount_point()
+
+    def _generate_mount_point(self):
+        result = "/mnt/anaconda-sources/{}-{:0>4}".format(self.type.value.lower(), self.counter)
```
Please use the same path we are using now. That means ``/run/install/...``

VladimirSlavik replied:
Done: Now paths are like /run/install/source-TYPE-NNNN.

jkonecny12 commented on pyanaconda/modules/payloads/source/source_base.py:
```diff
@@ -84,3 +89,43 @@ def tear_down_with_tasks(self):
+        self._counter_value = MountingSourceBase.counter
+        self._mount_point = "/mnt/anaconda-sources/{}-{:0>4}".format(self.type.value.lower(),
+                                                                     self.counter)
+        MountingSourceBase.counter = MountingSourceBase.counter + 1
```
I think that we want to store the id but not here but in the base class. It could be usable for plenty of other things (including logging) in the future and it would be great to have ID for every source. What do you think?

VladimirSlavik replied:
Do we want to decide that now? I agree keeping an ID would be somehow "proper", but  at this moment it is not really needed, and the mount point acts as a kind of ID, even if that is rather crude.

poncovka replied:
If we want to have ID for every source, than it is not definitely going to be `counter_value` in `MountingSourceBase`, so please remove that attribute. We can easily modify the code for generating mount points with this future ID.

VladimirSlavik replied:
Removed already - somehow GitHub keeps showing the old diff even if the other lines changed :(

In terms of keeping the ID, leaving that open for the future.

jkonecny12 commented:
Also I'm thinking that we should have a separate test file for base source implementation. To test this code separated from an existing sources.
Right now I'm thinking about the same for payloads.

VladimirSlavik commented:
@jkonecny12 yes, that's precisely what I am doing now...

jkonecny12 commented:
Yeah, I see that now. So I'll do the same for the payload base class. It would help me to test functionality for the `is_ready` solution.

VladimirSlavik commented:
Update:
- Addressed all clear feedback above
- Tests now exist
- Rebased

Left to discuss:
- Should the Harddrive inherit from the common mounting ancestors, and how? It's a poor match.

VladimirSlavik commented:
Update:
- Changed generation of mount points
- Harddrive is left for another PR
- Removed all commits with other concerns to separate PRs

poncovka commented:
> I think it would be better to move the `MountPointGenerator` to `core.utils` or somewhere like that. It seems to me like something general to be used. If that will be done than it should be renamed to something more general. Something like `UniquePathGenerator` and the INSTALLTREE should be removed. It also should have it's own tests to easily making sure that changing this class will still result in the same output used elsewhere.
> 
> `MountPointGenerator` than could be specialization of this class if needed.
> 
> Do you agree with me @poncovka ?

I would prefer to have as much code in the modules as possible. Also, I don't really see where else we could use that.

poncovka commented on tests/nosetests/pyanaconda_tests/module_source_base_test.py:
```diff
@@ -0,0 +1,122 @@
+        return []
+
+
+class DummySetUpMountTaskSubclass(SetUpMountTask):
```
In general, I don't think it is necessary to write tests with dummy implementations. It is enough the test one of the real implementations.

poncovka commented on pyanaconda/modules/payloads/source/source_base.py:
```diff
@@ -84,3 +91,43 @@ def tear_down_with_tasks(self):
+        :rtype: bool
+        """
+        res = os.path.ismount(self._mount_point)
+        log.debug("Source is set to %s ready state", res)
```
@jkonecny12, isn't this message a leftover from debugging? I would expect the payload to log this kind of messages when it has sources in invalid state. Otherwise, it is not really useful.

VladimirSlavik replied:
Removed.

pep8speaks commented:
Hello @VladimirSlavik! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:























* In the file [`tests/nosetests/pyanaconda_tests/module_source_base_test.py`](https://github.com/rhinstaller/anaconda/blob/9a746786325aee22d566597130e9c6b49214075f/tests/nosetests/pyanaconda_tests/module_source_base_test.py):

> [Line 94:100](https://github.com/rhinstaller/anaconda/blob/9a746786325aee22d566597130e9c6b49214075f/tests/nosetests/pyanaconda_tests/module_source_base_test.py#L94): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 99 characters)
> [Line 102:100](https://github.com/rhinstaller/anaconda/blob/9a746786325aee22d566597130e9c6b49214075f/tests/nosetests/pyanaconda_tests/module_source_base_test.py#L102): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 99 characters)

VladimirSlavik commented on pyanaconda/modules/payloads/source/mount_tasks.py:
```diff
@@ -0,0 +1,77 @@
+class SetUpMountTask(Task, ABC):
+    """Abstract base class for set up tasks that need mounting.
+
+    Subclasses must implement _do_mount().
```
Done!

VladimirSlavik commented:
Update: Removed some of the cruft you found in last round of reviews.

VladimirSlavik commented:
> I would like to see separate tests just for `MountPointGenerator` but I'll approve even this variant.

I'm thinking tests for the related "utils and stuff" would be my next PR after the harddrive source changes...
